### ISSUE TITLE
TypeScript 3.4 meta-issue

### ISSUE THREAD

**Issue by DanielRosenwasser** [babel/babel#9502]
TypeScript 3.4 will be introducing a number of syntactic additions in the type space as well as the expression space.

* [ ] The `readonly` type modifier for array types and tuple types. (#9503)
    * https://github.com/Microsoft/TypeScript/pull/29435
* [ ] Constant contexts (a.k.a. `as const`)
    * https://github.com/Microsoft/TypeScript/pull/29510
    * This seems to already work?

Our target date for the release of TypeScript 3.4 is around the end of March 2019.

**Comment by danez** [comment]
I tried parsing this `let x = 10 as const;` and it doesn't work afaics. How did you make it work?

```
Unexpected token (1:14)

> 1 | let x = 10 as const;
                              ^
```

**Comment by DanielRosenwasser** [comment]
I might be right - the online repl seems like it doesn't work with TypeScript for some reason.

**Comment by nicolo-ribaudo** [comment]
Doesn't the "Add plugin" button in the plugins section of the sidebar work for you?

**Comment by DanielRosenwasser** [comment]
Nope :(

> `Plugin @babel/plugin-transform-typescript could not be loaded`

**Comment by DanielRosenwasser** [comment]
Not even the preset works.

**Comment by danez** [comment]
@DanielRosenwasser should we wait with this changes till typescript 3.4 is out/close to being released?

**Comment by Jessidhia** [comment]
My intuition is that as soon as the `3.4.0` version is released it should be safe to release. TypeScript typically doesn't declare `.0` versions stable, though, they refer to them as "RC"; but I don't think there's any precedent for them breaking grammar or emit semantics after having released a `.0` version.

`@typescript-eslint/parser` always waits until the `.1` version (the one that gets the "stable version released" publicity) is out ~~which tbh is rather inconvenient~~.

---

@DanielRosenwasser would I be correct in my understanding that the `.0` version is intended to be supported by tooling and typing authors, and the "delay" before making it "stable" is to give some time for the ecosystem to catch up? Or should it just be treated like an oddly-named beta?

**Comment by DanielRosenwasser** [comment]
Well there's typically about 2 weeks in between an RC and a release. I would just wait until TS 3.4 goes out to publish anything with this functionality. If you want to let it "bake" in nightly releases, that's probably a good idea and you can pull that in.

**Comment by nicolo-ribaudo** [comment]
We usually don't do nightly releases, we just wait to merge PRs if needed since it is much simpler :laughing:

**Comment by ulrichb** [comment]
So, is this released with `plugin-transform-typescript` 7.4.0? ... I get "SyntaxError: Unexpected token" for both, readonly tuples and `as const` literals...

**Comment by ulrichb** [comment]
Okay, forget it. I know why, but I had an outdated `@babel/parser` package (< 7.4).
